File: lectern/coverage.py
from __future__ import annotations

from collections import Counter
from typing import Any, Dict, List, Set


def normalize_positive_int(value: Any) -> int | None:
    if isinstance(value, bool):
        return None
    if isinstance(value, int):
        return value if value > 0 else None
    if isinstance(value, float):
        int_value = int(value)
        return int_value if int_value > 0 else None
    if isinstance(value, str):
        stripped = value.strip()
        if stripped.isdigit():
            int_value = int(stripped)
            return int_value if int_value > 0 else None
    return None


def normalize_page_references(value: Any) -> List[int]:
    if value is None:
        return []
    if isinstance(value, (int, float, str)):
        normalized = normalize_positive_int(value)
        return [normalized] if normalized is not None else []

    refs: List[int] = []
    if isinstance(value, list):
        for item in value:
            normalized = normalize_positive_int(item)
            if normalized is not None and normalized not in refs:
                refs.append(normalized)
    return refs


def normalize_string_list(value: Any) -> List[str]:
    if value is None:
        return []
    if isinstance(value, str):
        items = [segment.strip() for segment in value.split(",")]
    elif isinstance(value, list):
        items = [str(item).strip() for item in value]
    else:
        return []
    return [item for item in items if item]


def get_card_page_references(card: Dict[str, Any]) -> List[int]:
    source_pages = normalize_page_references(card.get("source_pages"))
    if source_pages:
        return source_pages

    slide_number = normalize_positive_int(card.get("slide_number"))
    return [slide_number] if slide_number is not None else []


def get_card_concept_ids(card: Dict[str, Any]) -> List[str]:
    return normalize_string_list(card.get("concept_ids"))


def normalize_relation_key(value: Any) -> str:
    if not isinstance(value, str):
        return ""
    parts = [segment.strip() for segment in value.split("|", 2)]
    if len(parts) != 3 or not all(parts):
        return ""
    return "|".join(parts)


def make_relation_key(source: Any, rel_type: Any, target: Any) -> str:
    return normalize_relation_key(
        f"{str(source or '').strip()}|{str(rel_type or '').strip()}|{str(target or '').strip()}"
    )


def get_card_relation_keys(card: Dict[str, Any]) -> List[str]:
    relation_keys = normalize_string_list(card.get("relation_keys"))
    normalized = [normalize_relation_key(value) for value in relation_keys]
    return [value for value in normalized if value]


def build_coverage_catalog(
    concept_map: Dict[str, Any], total_pages: int
) -> Dict[str, Any]:
    concepts = concept_map.get("concepts") if isinstance(concept_map, dict) else []
    relations = concept_map.get("relations") if isinstance(concept_map, dict) else []

    concept_catalog: List[Dict[str, Any]] = []
    for concept in concepts if isinstance(concepts, list) else []:
        if not isinstance(concept, dict):
            continue
        concept_catalog.append(
            {
                "id": str(concept.get("id") or "").strip(),
                "name": str(concept.get("name") or "").strip(),
                "importance": str(concept.get("importance") or "medium"),
                "difficulty": str(concept.get("difficulty") or "intermediate"),
                "page_references": normalize_page_references(
                    concept.get("page_references")
                ),
            }
        )

    relation_catalog: List[Dict[str, Any]] = []
    for relation in relations if isinstance(relations, list) else []:
        if not isinstance(relation, dict):
            continue
        relation_catalog.append(
            {
                "source": str(relation.get("source") or "").strip(),
                "target": str(relation.get("target") or "").strip(),
                "type": str(relation.get("type") or "").strip(),
                "key": make_relation_key(
                    relation.get("source"),
                    relation.get("type"),
                    relation.get("target"),
                ),
                "page_references": normalize_page_references(
                    relation.get("page_references") or relation.get("page_reference")
                ),
            }
        )

    return {
        "total_pages": int(total_pages),
        "document_type": (
            concept_map.get("document_type") if isinstance(concept_map, dict) else None
        ),
        "concept_catalog": concept_catalog,
        "relation_catalog": relation_catalog,
    }


def compute_coverage_data(
    *,
    cards: List[Dict[str, Any]],
    concept_map: Dict[str, Any],
    total_pages: int,
) -> Dict[str, Any]:
    catalog = build_coverage_catalog(concept_map, total_pages)
    concept_catalog = catalog.get("concept_catalog", [])
    relation_catalog = catalog.get("relation_catalog", [])

    covered_pages: Set[int] = set()
    explicit_concept_ids: Set[str] = set()
    covered_concepts_by_page: Set[str] = set()
    explicit_relation_keys: Set[str] = set()
    covered_relations_by_page: Set[str] = set()
    cards_per_page: Counter[int] = Counter()

    for card in cards:
        if not isinstance(card, dict):
            continue
        page_refs = get_card_page_references(card)
        covered_pages.update(page_refs)
        for page in page_refs:
            cards_per_page[page] += 1
        explicit_concept_ids.update(get_card_concept_ids(card))
        explicit_relation_keys.update(get_card_relation_keys(card))

        if page_refs and concept_catalog:
            page_ref_set = set(page_refs)
            for concept in concept_catalog:
                concept_id = str(concept.get("id") or "").strip()
                concept_pages = set(
                    normalize_page_references(concept.get("page_references"))
                )
                if (
                    concept_id
                    and concept_pages
                    and page_ref_set.intersection(concept_pages)
                ):
                    covered_concepts_by_page.add(concept_id)
            for relation in relation_catalog:
                relation_key = str(relation.get("key") or "").strip()
                relation_pages = set(
                    normalize_page_references(relation.get("page_references"))
                )
                if (
                    relation_key
                    and relation_pages
                    and page_ref_set.intersection(relation_pages)
                ):
                    covered_relations_by_page.add(relation_key)

    covered_concept_ids_set = explicit_concept_ids.union(covered_concepts_by_page)
    covered_concept_ids = sorted(covered_concept_ids_set)
    concept_ids = [
        str(concept.get("id") or "").strip()
        for concept in concept_catalog
        if concept.get("id")
    ]
    high_priority_ids = [
        str(concept.get("id") or "").strip()
        for concept in concept_catalog
        if str(concept.get("importance") or "").strip() == "high" and concept.get("id")
    ]
    relation_keys = [
        str(relation.get("key") or "").strip()
        for relation in relation_catalog
        if relation.get("key")
    ]
    covered_relation_keys_set = explicit_relation_keys.union(covered_relations_by_page)

    uncovered_pages = [
        page for page in range(1, int(total_pages) + 1) if page not in covered_pages
    ]
    uncovered_concepts = [
        {
            "id": str(concept.get("id") or "").strip(),
            "name": str(concept.get("name") or "").strip(),
            "importance": str(concept.get("importance") or "medium"),
            "page_references": normalize_page_references(
                concept.get("page_references")
            ),
        }
        for concept in concept_catalog
        if str(concept.get("id") or "").strip() not in covered_concept_ids_set
    ]
    missing_high_priority = [
        concept for concept in uncovered_concepts if concept.get("importance") == "high"
    ]
    uncovered_relations = [
        {
            "key": str(relation.get("key") or "").strip(),
            "source": str(relation.get("source") or "").strip(),
            "target": str(relation.get("target") or "").strip(),
            "type": str(relation.get("type") or "").strip(),
            "page_references": normalize_page_references(
                relation.get("page_references")
            ),
        }
        for relation in relation_catalog
        if str(relation.get("key") or "").strip() not in covered_relation_keys_set
    ]
    saturated_pages = [
        page for page, count in sorted(cards_per_page.items()) if count > 2
    ]

    return {
        **catalog,
        "covered_pages": sorted(covered_pages),
        "uncovered_pages": uncovered_pages,
        "covered_page_count": len(covered_pages),
        "page_coverage_pct": (
            round((len(covered_pages) / total_pages) * 100) if total_pages > 0 else 0
        ),
        "saturated_pages": saturated_pages,
        "explicit_concept_count": len(explicit_concept_ids),
        "explicit_concept_coverage_pct": (
            round((len(explicit_concept_ids) / len(concept_ids)) * 100)
            if concept_ids
            else 0
        ),
        "covered_concept_ids": covered_concept_ids,
        "covered_concept_count": len(covered_concept_ids),
        "total_concepts": len(concept_ids),
        "concept_coverage_pct": (
            round((len(covered_concept_ids) / len(concept_ids)) * 100)
            if concept_ids
            else 0
        ),
        "explicit_relation_count": len(explicit_relation_keys),
        "covered_relation_count": len(covered_relation_keys_set),
        "total_relations": len(relation_keys),
        "relation_coverage_pct": (
            round((len(covered_relation_keys_set) / len(relation_keys)) * 100)
            if relation_keys
            else 0
        ),
        "high_priority_total": len(high_priority_ids),
        "high_priority_covered": len(
            [cid for cid in high_priority_ids if cid in covered_concept_ids_set]
        ),
        "missing_high_priority": missing_high_priority,
        "uncovered_concepts": uncovered_concepts,
        "uncovered_relations": uncovered_relations,
    }


def build_generation_gap_text(coverage_data: Dict[str, Any]) -> str:
    uncovered_pages = coverage_data.get("uncovered_pages") or []
    uncovered_concepts = coverage_data.get("uncovered_concepts") or []
    missing_high_priority = coverage_data.get("missing_high_priority") or []
    uncovered_relations = coverage_data.get("uncovered_relations") or []
    saturated_pages = coverage_data.get("saturated_pages") or []

    lines = [
        "- COVERAGE LEDGER:",
        f"  - Pages covered: {coverage_data.get('covered_page_count', 0)}/{coverage_data.get('total_pages', 0)}.",
        f"  - Concepts covered: {coverage_data.get('covered_concept_count', 0)}/{coverage_data.get('total_concepts', 0)} ({coverage_data.get('explicit_concept_count', 0)} explicit).",
        f"  - Relations covered: {coverage_data.get('covered_relation_count', 0)}/{coverage_data.get('total_relations', 0)} ({coverage_data.get('explicit_relation_count', 0)} explicit).",
    ]

    if uncovered_pages:
        preview = ", ".join(str(page) for page in uncovered_pages[:15])
        suffix = "..." if len(uncovered_pages) > 15 else ""
        lines.append(f"  - Prioritize uncovered pages: {preview}{suffix}")

    if missing_high_priority:
        preview = ", ".join(
            f"{item.get('name') or item.get('id')}@{','.join(str(p) for p in item.get('page_references') or []) or '?'}"
            for item in missing_high_priority[:8]
        )
        suffix = "..." if len(missing_high_priority) > 8 else ""
        lines.append(f"  - Missing HIGH priority concepts: {preview}{suffix}")
    elif uncovered_concepts:
        preview = ", ".join(
            f"{item.get('name') or item.get('id')}@{','.join(str(p) for p in item.get('page_references') or []) or '?'}"
            for item in uncovered_concepts[:8]
        )
        suffix = "..." if len(uncovered_concepts) > 8 else ""
        lines.append(f"  - Remaining concepts: {preview}{suffix}")

    if uncovered_relations:
        preview = ", ".join(
            f"{item.get('source')}|{item.get('type')}|{item.get('target')}"
            for item in uncovered_relations[:6]
        )
        suffix = "..." if len(uncovered_relations) > 6 else ""
        lines.append(f"  - Missing relations: {preview}{suffix}")

    if saturated_pages:
        preview = ", ".join(str(page) for page in saturated_pages[:8])
        suffix = "..." if len(saturated_pages) > 8 else ""
        lines.append(f"  - Over-covered pages to deprioritize: {preview}{suffix}")

    return "\n".join(lines) + "\n"


def build_reflection_gap_text(coverage_data: Dict[str, Any]) -> str:
    missing_high_priority = coverage_data.get("missing_high_priority") or []
    uncovered_pages = coverage_data.get("uncovered_pages") or []
    uncovered_relations = coverage_data.get("uncovered_relations") or []
    saturated_pages = coverage_data.get("saturated_pages") or []

    lines = [
        "Coverage audit:",
        f"- Page coverage: {coverage_data.get('covered_page_count', 0)}/{coverage_data.get('total_pages', 0)}.",
        f"- Concept coverage: {coverage_data.get('covered_concept_count', 0)}/{coverage_data.get('total_concepts', 0)} ({coverage_data.get('explicit_concept_count', 0)} explicit).",
        f"- Relation coverage: {coverage_data.get('covered_relation_count', 0)}/{coverage_data.get('total_relations', 0)} ({coverage_data.get('explicit_relation_count', 0)} explicit).",
        f"- High-priority concept coverage: {coverage_data.get('high_priority_covered', 0)}/{coverage_data.get('high_priority_total', 0)}.",
    ]
    if uncovered_pages:
        lines.append(
            "- Uncovered pages: "
            + ", ".join(str(page) for page in uncovered_pages[:15])
            + ("..." if len(uncovered_pages) > 15 else "")
        )
    if missing_high_priority:
        lines.append(
            "- Missing high-priority concepts: "
            + ", ".join(
                str(item.get("name") or item.get("id"))
                for item in missing_high_priority[:10]
            )
            + ("..." if len(missing_high_priority) > 10 else "")
        )
    if uncovered_relations:
        lines.append(
            "- Missing relations: "
            + ", ".join(
                f"{item.get('source')}|{item.get('type')}|{item.get('target')}"
                for item in uncovered_relations[:8]
            )
            + ("..." if len(uncovered_relations) > 8 else "")
        )
    if saturated_pages:
        lines.append(
            "- Saturated pages to thin out: "
            + ", ".join(str(page) for page in saturated_pages[:10])
            + ("..." if len(saturated_pages) > 10 else "")
        )
    return "\n".join(lines)

File: src/engine/coverage.ts
/**
 * Coverage ledger — tracks which pages / concepts / relations of the concept
 * map are already covered by generated cards, renders gap text that steers
 * the model toward uncovered material, and decides when coverage is
 * sufficient to stop.
 */

import {
  COVERAGE_MIN_CONCEPT_PERCENT,
  COVERAGE_MIN_PAGE_PERCENT,
  COVERAGE_MIN_RELATION_PERCENT,
  SATURATION_CARDS_PER_PAGE,
} from './config'
import { normalizeRelationKey } from './quality'
import type { Card, ConceptMap, CoverageCatalog, CoverageData } from './types'
import { relationKeyOf } from './types'

// ---------------------------------------------------------------------------
// Local extension of the shared catalog type
// ---------------------------------------------------------------------------

/**
 * The shared CoverageCatalog does not carry relation page references, but the
 * Python original infers relation coverage from page overlap. We keep that
 * behavior by extending the contract type locally (structurally compatible —
 * every EngineCoverageCatalog is a valid CoverageCatalog).
 */
export interface EngineCoverageCatalog extends CoverageCatalog {
  /** page number -> relation keys referenced on that page */
  relationsByPage: Map<number, string[]>
}

const isEngineCatalog = (catalog: CoverageCatalog): catalog is EngineCoverageCatalog =>
  (catalog as Partial<EngineCoverageCatalog>).relationsByPage instanceof Map

// ---------------------------------------------------------------------------
// Normalization helpers (ports of coverage.py normalize_*)
// ---------------------------------------------------------------------------

/** Positive-int normalization: truncate, keep > 0, dedupe preserving order. */
const normalizePages = (pages: readonly number[]): number[] => {
  const result: number[] = []
  const seen = new Set<number>()
  for (const raw of pages) {
    if (!Number.isFinite(raw)) continue
    const page = Math.trunc(raw)
    if (page > 0 && !seen.has(page)) {
      result.push(page)
      seen.add(page)
    }
  }
  return result
}

/** Port of get_card_page_references: sourcePages, else slideNumber fallback. */
const cardPageRefs = (card: Card): number[] => {
  const pages = normalizePages(card.sourcePages)
  if (pages.length > 0) return pages
  if (card.slideNumber !== undefined && Number.isFinite(card.slideNumber)) {
    const slide = Math.trunc(card.slideNumber)
    if (slide > 0) return [slide]
  }
  return []
}

const cardConceptIds = (card: Card): string[] =>
  card.conceptIds.map((id) => id.trim()).filter((id) => id !== '')

const cardRelationKeys = (card: Card): string[] =>
  card.relationKeys.map((key) => normalizeRelationKey(key)).filter((key) => key !== '')

// ---------------------------------------------------------------------------
// Catalog building (port of build_coverage_catalog)
// ---------------------------------------------------------------------------

export function buildCoverageCatalog(conceptMap: ConceptMap): EngineCoverageCatalog {
  const conceptIds = new Set<string>()
  const highPriorityIds = new Set<string>()
  const conceptNames = new Map<string, string>()
  const pagesByConcept = new Map<string, number[]>()
  const conceptsByPage = new Map<number, string[]>()

  for (const concept of conceptMap.concepts) {
    const id = concept.id.trim()
    if (id === '') continue
    conceptIds.add(id)
    if (concept.importance === 'high') highPriorityIds.add(id)
    conceptNames.set(id, concept.name.trim())

    const pages = normalizePages(concept.pageReferences)
    const existing = pagesByConcept.get(id)
    if (existing === undefined) {
      pagesByConcept.set(id, pages)
    } else {
      for (const page of pages) {
        if (!existing.includes(page)) existing.push(page)
      }
    }
    for (const page of pages) {
      const onPage = conceptsByPage.get(page)
      if (onPage === undefined) {
        conceptsByPage.set(page, [id])
      } else if (!onPage.includes(id)) {
        onPage.push(id)
      }
    }
  }

  const relationKeys = new Set<string>()
  const relationsByPage = new Map<number, string[]>()
  for (const relation of conceptMap.relations) {
    const key = normalizeRelationKey(
      relationKeyOf({
        source: relation.source.trim(),
        type: relation.type.trim(),
        target: relation.target.trim(),
      }),
    )
    if (key === '') continue
    relationKeys.add(key)
    for (const page of normalizePages(relation.pageReferences)) {
      const onPage = relationsByPage.get(page)
      if (onPage === undefined) {
        relationsByPage.set(page, [key])
      } else if (!onPage.includes(key)) {
        onPage.push(key)
      }
    }
  }

  return {
    conceptIds,
    highPriorityIds,
    relationKeys,
    conceptsByPage,
    pagesByConcept,
    conceptNames,
    pageCount: conceptMap.pageCount,
    relationsByPage,
  }
}

// ---------------------------------------------------------------------------
// Coverage computation (port of compute_coverage_data)
// ---------------------------------------------------------------------------

const sortedNumbers = (values: Iterable<number>): number[] => [...values].sort((a, b) => a - b)

const percent = (covered: number, total: number): number =>
  total > 0 ? Math.round((covered / total) * 100) : 0

export function computeCoverageData(catalog: CoverageCatalog, cards: Card[]): CoverageData {
  const relationsByPage = isEngineCatalog(catalog)
    ? catalog.relationsByPage
    : new Map<number, string[]>()

  const coveredPages = new Set<number>()
  const explicitConceptIds = new Set<string>()
  const conceptsCoveredByPage = new Set<string>()
  const explicitRelationKeys = new Set<string>()
  const relationsCoveredByPage = new Set<string>()
  const cardsPerPage = new Map<number, number>()

  for (const card of cards) {
    const pages = cardPageRefs(card)
    for (const page of pages) {
      coveredPages.add(page)
      cardsPerPage.set(page, (cardsPerPage.get(page) ?? 0) + 1)
      for (const conceptId of catalog.conceptsByPage.get(page) ?? []) {
        conceptsCoveredByPage.add(conceptId)
      }
      for (const relationKey of relationsByPage.get(page) ?? []) {
        relationsCoveredByPage.add(relationKey)
      }
    }
    for (const conceptId of cardConceptIds(card)) explicitConceptIds.add(conceptId)
    for (const relationKey of cardRelationKeys(card)) explicitRelationKeys.add(relationKey)
  }

  const coveredConceptUnion = new Set<string>(explicitConceptIds)
  for (const conceptId of conceptsCoveredByPage) coveredConceptUnion.add(conceptId)

  const coveredRelationUnion = new Set<string>(explicitRelationKeys)
  for (const relationKey of relationsCoveredByPage) coveredRelationUnion.add(relationKey)

  const uncoveredPages: number[] = []
  for (let page = 1; page <= catalog.pageCount; page += 1) {
    if (!coveredPages.has(page)) uncoveredPages.push(page)
  }

  const missingHighPriority = [...catalog.highPriorityIds].filter(
    (id) => !coveredConceptUnion.has(id),
  )

  const saturatedPages = sortedNumbers(cardsPerPage.keys()).filter(
    (page) => (cardsPerPage.get(page) ?? 0) > SATURATION_CARDS_PER_PAGE,
  )

  const cardsPerPageRecord: Record<number, number> = {}
  for (const [page, count] of cardsPerPage) cardsPerPageRecord[page] = count

  const totalConcepts = catalog.conceptIds.size
  const totalRelations = catalog.relationKeys.size

  return {
    pageCount: catalog.pageCount,
    coveredPages: sortedNumbers(coveredPages),
    uncoveredPages,
    pageCoveragePercent: percent(coveredPages.size, catalog.pageCount),
    coveredConceptIds: [...explicitConceptIds].sort(),
    inferredConceptIds: [...conceptsCoveredByPage]
      .filter((id) => !explicitConceptIds.has(id))
      .sort(),
    conceptCoveragePercent: percent(explicitConceptIds.size, totalConcepts),
    effectiveConceptCoveragePercent: percent(coveredConceptUnion.size, totalConcepts),
    coveredRelationKeys: [...coveredRelationUnion].sort(),
    // When the catalog has no relations there is nothing left to cover, so
    // report 100 (Python reported 0 but special-cased "no relations" in the
    // sufficiency check; CoverageData alone must carry that information here).
    relationCoveragePercent:
      totalRelations > 0 ? percent(coveredRelationUnion.size, totalRelations) : 100,
    missingHighPriority,
    saturatedPages,
    cardsPerPage: cardsPerPageRecord,
  }
}

// ---------------------------------------------------------------------------
// Gap text (ports of build_generation_gap_text / build_reflection_gap_text)
// ---------------------------------------------------------------------------

const preview = <T>(items: readonly T[], limit: number, render: (item: T) => string): string =>
  items.slice(0, limit).map(render).join(', ') + (items.length > limit ? '...' : '')

const describeConcept = (catalog: CoverageCatalog, id: string): string => {
  const name = catalog.conceptNames.get(id)
  const pages = catalog.pagesByConcept.get(id) ?? []
  const pageText = pages.length > 0 ? pages.join(',') : '?'
  return `${name !== undefined && name !== '' ? name : id}@${pageText}`
}

interface GapSummary {
  coveredConceptCount: number
  explicitConceptCount: number
  coveredRelationCount: number
  uncoveredConceptIds: string[]
  uncoveredRelationKeys: string[]
}

const summarizeGaps = (catalog: CoverageCatalog, coverage: CoverageData): GapSummary => {
  const coveredConcepts = new Set<string>([
    ...coverage.coveredConceptIds,
    ...coverage.inferredConceptIds,
  ])
  const coveredRelations = new Set<string>(coverage.coveredRelationKeys)
  return {
    coveredConceptCount: coveredConcepts.size,
    explicitConceptCount: coverage.coveredConceptIds.length,
    coveredRelationCount: coverage.coveredRelationKeys.length,
    uncoveredConceptIds: [...catalog.conceptIds].filter((id) => !coveredConcepts.has(id)),
    uncoveredRelationKeys: [...catalog.relationKeys].filter((key) => !coveredRelations.has(key)),
  }
}

export function buildGenerationGapText(catalog: CoverageCatalog, coverage: CoverageData): string {
  const gaps = summarizeGaps(catalog, coverage)

  const lines = [
    '- COVERAGE LEDGER:',
    `  - Pages covered: ${coverage.coveredPages.length}/${catalog.pageCount}.`,
    `  - Concepts covered: ${gaps.coveredConceptCount}/${catalog.conceptIds.size} (${gaps.explicitConceptCount} explicit).`,
    `  - Relations covered: ${gaps.coveredRelationCount}/${catalog.relationKeys.size}.`,
  ]

  if (coverage.uncoveredPages.length > 0) {
    lines.push(`  - Prioritize uncovered pages: ${preview(coverage.uncoveredPages, 15, String)}`)
  }

  if (coverage.missingHighPriority.length > 0) {
    lines.push(
      `  - Missing HIGH priority concepts: ${preview(coverage.missingHighPriority, 8, (id) =>
        describeConcept(catalog, id),
      )}`,
    )
  } else if (gaps.uncoveredConceptIds.length > 0) {
    lines.push(
      `  - Remaining concepts: ${preview(gaps.uncoveredConceptIds, 8, (id) =>
        describeConcept(catalog, id),
      )}`,
    )
  }

  if (gaps.uncoveredRelationKeys.length > 0) {
    lines.push(`  - Missing relations: ${preview(gaps.uncoveredRelationKeys, 6, (key) => key)}`)
  }

  if (coverage.saturatedPages.length > 0) {
    lines.push(
      `  - Over-covered pages to deprioritize: ${preview(coverage.saturatedPages, 8, String)}`,
    )
  }

  return lines.join('\n') + '\n'
}

export function buildReflectionGapText(catalog: CoverageCatalog, coverage: CoverageData): string {
  const gaps = summarizeGaps(catalog, coverage)
  const highPriorityTotal = catalog.highPriorityIds.size
  const highPriorityCovered = highPriorityTotal - coverage.missingHighPriority.length

  const lines = [
    'Coverage audit:',
    `- Page coverage: ${coverage.coveredPages.length}/${catalog.pageCount}.`,
    `- Concept coverage: ${gaps.coveredConceptCount}/${catalog.conceptIds.size} (${gaps.explicitConceptCount} explicit).`,
    `- Relation coverage: ${gaps.coveredRelationCount}/${catalog.relationKeys.size}.`,
    `- High-priority concept coverage: ${highPriorityCovered}/${highPriorityTotal}.`,
  ]

  if (coverage.uncoveredPages.length > 0) {
    lines.push(`- Uncovered pages: ${preview(coverage.uncoveredPages, 15, String)}`)
  }
  if (coverage.missingHighPriority.length > 0) {
    lines.push(
      `- Missing high-priority concepts: ${preview(coverage.missingHighPriority, 10, (id) => {
        const name = catalog.conceptNames.get(id)
        return name !== undefined && name !== '' ? name : id
      })}`,
    )
  }
  if (gaps.uncoveredRelationKeys.length > 0) {
    lines.push(`- Missing relations: ${preview(gaps.uncoveredRelationKeys, 8, (key) => key)}`)
  }
  if (coverage.saturatedPages.length > 0) {
    lines.push(`- Saturated pages to thin out: ${preview(coverage.saturatedPages, 10, String)}`)
  }

  return lines.join('\n')
}

// ---------------------------------------------------------------------------
// Sufficiency (port of generation_utils._coverage_is_sufficient)
// ---------------------------------------------------------------------------

export function isCoverageSufficient(coverage: CoverageData): boolean {
  const highPriorityOk = coverage.missingHighPriority.length === 0
  // relationCoveragePercent is 100 when the catalog has no relations, which
  // mirrors Python's "total_relations == 0 or relation_pct >= 50" check.
  const relationOk = coverage.relationCoveragePercent >= COVERAGE_MIN_RELATION_PERCENT
  return (
    highPriorityOk &&
    relationOk &&
    (coverage.conceptCoveragePercent >= COVERAGE_MIN_CONCEPT_PERCENT ||
      coverage.pageCoveragePercent >= COVERAGE_MIN_PAGE_PERCENT)
  )
}
